EA Recent discussions redesign

oetherington opened the pull request:
This PR implements the recent discussions redesign for the EA Forum. [Figma](https://www.figma.com/file/d09iVzNSpZp9GyWGcfkWmX/Q2-2023?node-id=1953%3A2947&amp;mode=dev).Most of the changes are done only on the frontend we switching out components in the mixed type feed for vew variants. The one exception to this is quick takes which required adding a new type to the mixed type feed on the backend to show the main quick take body along with comments on that quick take.In also strayed slightly away from the Figma when implementing the "More" button at the end of post bodies. The figma specified using an ellipsis here, but there's no straight forward way to do this in CSS as designed and it would require a slow/complicated/bug-prone JS solution. Instead I added a fade:

![](https://github.com/ForumMagnum/ForumMagnum/assets/5075628/304ae61c-e68c-42ce-96c0-c7f562db2c88)New posts:

![](https://github.com/ForumMagnum/ForumMagnum/assets/5075628/2f3c1e91-65cc-4b03-8127-d3d5254dfc37)Comments on posts:

![](https://github.com/ForumMagnum/ForumMagnum/assets/5075628/d7eb339f-ad55-4a71-afef-017c48e36317)New quick takes:

![](https://github.com/ForumMagnum/ForumMagnum/assets/5075628/fa0d3fcf-24c3-4cbe-87b8-05116354b2d6)Comments on quick takes:

![](https://github.com/ForumMagnum/ForumMagnum/assets/5075628/c4cad279-cb86-4a81-b984-23a8bc71fe08)New events:

![](https://github.com/ForumMagnum/ForumMagnum/assets/5075628/f2114662-74c1-4f0f-82f2-f7742bb4977f)Tag edits:

![](https://github.com/ForumMagnum/ForumMagnum/assets/5075628/4e7d578b-5943-45ce-8d45-dd8a4eac37ae)Tag comments (these weren't in the figma design so I just made them roughly match the other types - these are very rare anyway):

![](https://github.com/ForumMagnum/ForumMagnum/assets/5075628/2bcb4e9a-82b5-406a-8eb1-cf2f858ba094)I also updated the background and border colors of the digest subscription reminder to match the other items in the feed:

![](https://github.com/ForumMagnum/ForumMagnum/assets/5075628/97bd4e95-7e23-42b1-8080-00aa03c59077)



┆Issue is synchronized with this [Asana task](https://app.asana.com/0/1201302964208280/1205491853880838) by [Unito](https://www.unito.io)

oetherington commented on packages/lesswrong/components/ea-forum/EAPostMeta.tsx:
This is refactored out of `EAPostsItem` for use elsewhere

oetherington commented on packages/lesswrong/components/common/EAKarmaDisplay.tsx:
This is refactored out of `EAPostsItem` for use elsewhere

oetherington commented on packages/lesswrong/components/recentDiscussion/useRecentDiscussionThread.ts:
This is all refactored out of `RecentDiscussionThread`

oetherington commented on packages/lesswrong/server/resolvers/recentDiscussionFeed.ts:
```diff
@@ -67,6 +62,23 @@ defineFeedResolver<Date>({
       {[`tagRelevance.${EA_FORUM_TRANSLATION_TOPIC_ID}`]: {$exists: false}},
     ]};
 
+    const postSelector = {
```
The changes here are just to separate out shortforms from other types of posts so that we can return the `topLevelComment` with any shortform comments.

oetherington commented on packages/lesswrong/themes/userThemes/darkMode.ts:
```diff
@@ -46,7 +46,11 @@ export const invertedGreyscale = {
+const inverseGreyAlpha = (alpha: number) => {
+  const [r, g, b] = invertColor([1, 1, 1, 1]);
+  return `rgba(${zeroTo255(r)},${zeroTo255(g)},${zeroTo255(b)},${alpha})`;
+}
```
This change is needed because we apply some biasing when convert shades of grey into dark mode to make the contrast more attractive, however, the biasing isn't currently applied to this function so the colors it returns are slightly wrong.

oetherington commented:
I think this is ready for your eyes @agnesstenlund

agnesstenlund commented:
woah this looks really great 👏 

I'll make some small padding adjustments + see if I can find anything else I think should change 🙏 

Right now my only request is to show post hovers when hovering the titles up here (after `posted`)
![Screenshot 2023-09-15 at 16 52 12](https://github.com/ForumMagnum/ForumMagnum/assets/13235378/1ecf1f1c-833a-47c6-9ea7-b5cf8c3329c6)

agnesstenlund commented:
when you untick "show on the Frontpage" I would assume they shouldn't show up in this feed? Or do personal posts show up here?

 ![Screenshot 2023-09-15 at 17 25 54](https://github.com/ForumMagnum/ForumMagnum/assets/13235378/99bd77cd-07ca-42c0-afd2-8d11ff6d41f0)

agnesstenlund commented:
- [ ] small thing (which I also want to change for posts) but I think headings should have the same color as the title (right now they're slightly lighter grey)

![Screenshot 2023-09-15 at 17 32 01](https://github.com/ForumMagnum/ForumMagnum/assets/13235378/710dc624-ffc4-478a-9705-c00d903d4c0d)

agnesstenlund commented:
- [ ] Not a deal-breaker but would be nice if the "More" was next to the text when the text doesn't fill up the whole width
![Screenshot 2023-09-15 at 17 36 36](https://github.com/ForumMagnum/ForumMagnum/assets/13235378/71539755-6a87-4e84-b4e4-557e30cf9dab)

- [ ] Would be nice if links got darker on hover + especially that the visited link turned a different purple vs turn back to blue

https://github.com/ForumMagnum/ForumMagnum/assets/13235378/43fa2284-0a8e-4bec-a6d4-d2bf0998e23b

agnesstenlund commented:
- [ ] I think this one should have the blue comment bubble (even though it's a tag)
![Screenshot 2023-09-15 at 17 43 48](https://github.com/ForumMagnum/ForumMagnum/assets/13235378/8f5bebe3-b835-46ec-a8e7-3fc35dcf04de)

agnesstenlund commented:
- [ ] For [questions](https://cea-core.slack.com/archives/C03G0AQ925P/p1694792820212259?thread_ts=1694792751.034749&cid=C03G0AQ925P), lets use the Q icon after all

![image](https://github.com/ForumMagnum/ForumMagnum/assets/13235378/2c0039c1-8029-4e12-8bc6-697b4cfef6e3)

- [ ] Also include the Q and linkpost link icon next to the title like we do for the standard postItem 🙌 
![Screenshot 2023-09-15 at 17 54 16](https://github.com/ForumMagnum/ForumMagnum/assets/13235378/2400ae3e-7a6b-40a8-94a2-abbf8ac877b5)
![Screenshot 2023-09-15 at 17 54 13](https://github.com/ForumMagnum/ForumMagnum/assets/13235378/fc1fa078-ab4e-4333-9c9e-00e605a78abe)

oetherington commented:
@agnesstenlund Thanks! I've just addressed the comments above.

- I've just excluded personal blogposts from recent discussions. This is different to how it currently works, but I think this is probably just a bug at the moment.
- Where's the "Q" icon for questions? I looked in the figma but couldn't see any (sorry if I'm being blind!)
- Moving the "(More)" next to the text is really difficult for the same reason as using an "..." in the same place. I think we should probably leave it as-is for now, and then change this in the future if we think it's worth the time investment.

agnesstenlund commented:
Thank you!!

> Where's the "Q" icon for questions? I looked in the figma but couldn't see any (sorry if I'm being blind!)


sorry this isn't in figma 🙈 but I know we use that icon on the postitem when there is a question. But maybe that's not an icon but just a Q 

![Screenshot 2023-09-18 at 16 29 23](https://github.com/ForumMagnum/ForumMagnum/assets/13235378/ed30476a-abba-41fc-865b-4bc562647871)

> Moving the "(More)" next to the text is really difficult for the same reason as using an "..." in the same place. I think we should probably leave it as-is for now, and then change this in the future if we think it's worth the time investment.

makes sense, sounds like a good call

oetherington commented:
Oh, yeah - I did add the "Q" letter after the title but I thought you meant you wanted to use a different icon on the left here:
<img width="49" alt="Screenshot 2023-09-18 at 17 34 55" src="https://github.com/ForumMagnum/ForumMagnum/assets/5075628/c36f2b87-20cf-433b-aead-beec4ced6487">

I may have just misunderstood though. Thanks!

agnesstenlund commented:
my bad, I'm being vague–I was thinking like this 👇 
![Screenshot 2023-09-19 at 11 24 03](https://github.com/ForumMagnum/ForumMagnum/assets/13235378/3e78456b-9272-4034-8dd6-47c8907b7103)

agnesstenlund commented:
One last request–

I agree with the tradeoff of not putting (More) next to the text. Looking more at it, I do think it looks a little odd.

- [ ] I'd suggest we change both the recent discussion cards + the cards on topics pages to show "Continue reading" below the text, like this:

![Screenshot 2023-09-19 at 11 29 26](https://github.com/ForumMagnum/ForumMagnum/assets/13235378/16779078-6c3f-4afb-ad6d-7cacc6553623)

Right now cards on topic pages look like this, so we can use that same style but remove the ( ) and "2540 more words":
![Screenshot 2023-09-19 at 11 29 05](https://github.com/ForumMagnum/ForumMagnum/assets/13235378/46da494c-06c2-46ce-80bd-1222340c23a5)

would reduce top padding to 12px instead of 16px though
![Screenshot 2023-09-19 at 11 33 17](https://github.com/ForumMagnum/ForumMagnum/assets/13235378/5b9939e1-48d6-4156-b66f-c3f050b558e3)

agnesstenlund commented:
- [ ] Also found this–looks like there isn't enough padding before the comment
![Screenshot 2023-09-19 at 18 31 43](https://github.com/ForumMagnum/ForumMagnum/assets/13235378/1a16c49c-fce8-43fc-b0c1-f82a38c9327a)

agnesstenlund commented:
- [ ] Oh also, apart from being blue (which is correct) this should also actually be a comment bubble, not a tag icon
![Screenshot 2023-09-19 at 18 32 32](https://github.com/ForumMagnum/ForumMagnum/assets/13235378/2b184b97-cf50-4473-b8a1-bf2eee2cf258)

s-cheng commented:
@agnesstenlund I think the "see more" vs "continue reading" tells you whether the card will expand to show the rest of the text vs if it will link you to the post page? I currently find this distinction useful, though maybe we should change the copy since it's not totally clear

agnesstenlund commented:
I see, didn't realise. We can just leave the "see more" for now then, agree they should be distinct if they trigger different actions 👍

oetherington commented:
Just pushed some commits which should hopefully fix all the above issues 👍

agnesstenlund commented:
Great! All good on my end then 🚢

jpaddison3 commented:
I love this so much.

Here,

<img width="916" alt="image" src="https://github.com/ForumMagnum/ForumMagnum/assets/10352319/e033d719-46c3-4a84-ae9f-6625a35a406b">

I would like to see a ⬑ arrow. Currently it seems to me that Jeff is commenting directly on Jaimie's quick take, which is wrong and disorienting.

oetherington commented:
@jpaddison3 Should be fixed now

jpaddison3 commented:
These two should have collapsed into one

<img width="869" alt="image" src="https://github.com/ForumMagnum/ForumMagnum/assets/10352319/9ca941f0-927e-47e7-b74e-5b7c37b37aee">

jpaddison3 commented:
Can I take this annoying opportunity to suggest a change to the design? If the post / quick take / tag description is less than ~ 10(?) lines, can it expand in-situ like it does today instead of going to the post? I agree large posts should probably go to the post.

Actually my heuristic might be:

Quick take — in situ
Post — if ≤ 10 lines, in situ
Tags — "

Anyway, @agnesstenlund, what'd'you think?

oetherington commented:
@jpaddison3 Quick take responses should now merge

jpaddison3 commented:
This guy is misaligned

![image](https://github.com/ForumMagnum/ForumMagnum/assets/10352319/2cb55577-3a0c-4565-8a93-262ef50b8d46)

jpaddison3 commented:
I've gotten a few

```
2023-09-21T20:08:33.973Z: [GraphQLError: invalid input syntax for type timestamp with time zone: "0NaN-NaN-NaNTNaN:NaN:NaN.NaN+NaN:NaN"] {
```

Errors on the server that prevents me from scrolling down through recent discussion.

jpaddison3 commented on packages/lesswrong/components/tagging/TagTooltipWrapper.tsx:
```diff
@@ -0,0 +1,46 @@
+import { Components, registerComponent } from "../../lib/vulcan-lib";
+import { useHover } from "../common/withHover";
+
+const TagTooltipWrapper = ({tag, tagRel, popperCard, As = "div", children}: {
```
`As = "div"`

That works??

oetherington replied:
Yep!

jpaddison3 reviewed (approved):
ASTLAC

agnesstenlund commented:
> If the post / quick take / tag description is less than ~ 10(?) lines, can it expand in-situ like it does today instead of going to the post? 

Yeah I think that's sensible 👍  was thinking at first we should use different language here so you know what to expect, but I actually think "Continue reading" is fine for both, and you'd probably expect that to expand (and get a bit disappointed when you're launched to a new page, but it's fine)

oetherington commented on packages/lesswrong/components/common/excerpts/CommentExcerpt.tsx:
```diff
@@ -0,0 +1,40 @@
+import { Components, registerComponent } from "../../../lib/vulcan-lib";
+import { commentGetPageUrlFromIds } from "../../../lib/collections/comments/helpers";
+
+const CommentExcerpt = ({comment, lines = 3, className}: {
```
Maybe - I think it depends on how much we like this once it's released

oetherington commented on packages/lesswrong/components/recentDiscussion/useRecentDiscussionThread.ts:
```diff
@@ -0,0 +1,69 @@
+
+  const lastVisitedAt = markedAsVisitedAt || post.lastVisitedAt
+
+  // TODO verify whether/how this should be interacting with afCommentCount
```
This comment is actually just moved from when this was refactored out of `RecentDiscussionThread`. I think there's a good chance it's out of date, but since it doesn't actually pertain to us I didn't want to remove it.

agnesstenlund commented:
Sorry, noticed one more thing:

- [ ] We should grey titles you've already read


Also, I got some questions from a user that I wasn't sure of the answer to:

1. If two people comment at the same time, which name do we show on the event? Perhaps we should show both? `Agnes and Ollie commented on` 
2. How many comments do we merge into one event in general? As in, if I comment 10 minutes after you comment, will that become the same event on the timeline?
3. We don't exclude stuff you've already read, right? So if you've already read a post or comment, it'd still show up in Recent discussion?

oetherington commented:
Above comments should now be addressed. I couldn't repro the error with the NaN timestamps, but I'll keep and eye out for it (my guess is it was probably related to a specific post that's no longer in the feed).

RE Agnes' questions:
 1) Currently we show the name of the newest comment - we could add the others, but I worry it could get too long?
 2) We fetch the comments based on the time they were made rather than by a count so all comments on the same post will be merged into a single recent discussions items - we fetch the last 5 by default
 3) Yep - that's correct

agnesstenlund commented:
thanks!!

> Currently we show the name of the newest comment - we could add the others, but I worry it could get too long?

yeah that seems right–it could in theory be up to 5 names? Or more?

agnesstenlund commented:
are all 5 expanded or only the newest comment? If they're all expanded it may actually be nice to say "Agnes, Ollie, Sarah, Will and JP commented on x" but otherwise seems good to only show the most recent one. Probably not a big deal either way

oetherington commented:
I'm pretty sure it only expands the most recent comment so I'm going to leave it as it is for now and merge into the feature branch on the beta site. We can make a separate PR if we want to change it in the future though 👍
